pkg/provisioningv2/rke2/planner/initnode.go

```diff
@@ -0,0 +1,255 @@
+	if _, ok := machine.Labels[InitNodeLabel]; !ok {
+		return nil
+	}
+	machine = machine.DeepCopy()
+	delete(machine.Labels, InitNodeLabel)
+	_, err := p.machines.Update(machine)
+	if err != nil {
+		return err
+	}
+	// We've changed state, so let the caches sync up again
+	return generic.ErrSkip
+}
+
+// setInitNodeMark sets the init node label on the given machine and updates the machine directly against the api
+// server. It returns the modified/updated machine object
+func (p *Planner) setInitNodeMark(machine *capi.Machine) (*capi.Machine, error) {
+	if machine.Labels[InitNodeLabel] == "true" {
+		return machine, nil
+	}
+	machine = machine.DeepCopy()
+	if machine.Labels == nil {
+		machine.Labels = map[string]string{}
+	}
+	machine.Labels[InitNodeLabel] = "true"
+	if _, err := p.machines.Update(machine); err != nil {
+		return nil, err
+	}
+	// We've changed state, so let the caches sync up again
+	return nil, generic.ErrSkip
+}
+
+// findAndDesignateFixedInitNode is used for rancherd where an exact machine (determined by labeling the
+// rkecontrolplane object) is desired to be the init node
+func (p *Planner) findAndDesignateFixedInitNode(rkeControlPlane *rkev1.RKEControlPlane, plan *plan.Plan) (bool, string, error) {
+	logrus.Debugf("rkecluster %s/%s: finding and designating fixed init node", rkeControlPlane.Namespace, rkeControlPlane.Spec.ClusterName)
+	fixedMachineID := rkeControlPlane.Labels[InitNodeMachineIDLabel]
+	if fixedMachineID == "" {
+		return false, "", fmt.Errorf("fixed machine ID label did not exist on rkecontrolplane")
+	}
+	entries := collect(plan, func(machine *capi.Machine) bool {
+		return machine.Labels[MachineIDLabel] == fixedMachineID
+	})
+	if len(entries) > 1 {
+		return false, "", fmt.Errorf("multiple machines found with identical machine ID label %s=%s", MachineIDLabel, fixedMachineID)
+	} else if len(entries) == 0 {
+		return false, "", fmt.Errorf("fixed machine with ID %s not found", fixedMachineID)
+	}
+	if rkeControlPlane.Labels[InitNodeMachineIDDoneLabel] == "" {
+		logrus.Debugf("rkecluster %s/%s: setting designated init node to fixedMachineID: %s", rkeControlPlane.Namespace, rkeControlPlane.Spec.ClusterName, fixedMachineID)
+		allInitNodes := collect(plan, isEtcd)
+		// clear all init node marks and return a generic.ErrSkip if we invalidated caches during clearing
+		cachesInvalidated := false
+		for _, entry := range allInitNodes {
+			if entry.Machine.Labels[MachineIDLabel] == fixedMachineID {
+				continue
+			}
+			err := p.clearInitNodeMark(entry.Machine)
+			if err != nil && !errors.Is(err, generic.ErrSkip) {
+				// if we received a strange error attempting to clear the init node mark
+				return false, "", err
+			} else if err != nil && errors.Is(err, generic.ErrSkip) {
+				cachesInvalidated = true
+			}
+		}
+		if cachesInvalidated {
+			return false, "", generic.ErrSkip
+		}
+		_, err := p.setInitNodeMark(entries[0].Machine)
+		if err != nil && !errors.Is(err, generic.ErrSkip) {
```

Comment:
Should we also "invalidate the caches" here and return the `generic.ErrSkip`?

Reply:
I don't think this is necessary for the `setInitNodeMark` because we are then modifying the rkecontrolplane object which only loosely relates to the machine we just marked as the init node, and we eventually return a `generic.ErrSkip` anyway to invalidate the caches later on.